apps/emqx_durable_storage/src/emqx_ds_storage_bitfield_lts.erl

```diff
@@ -0,0 +1,382 @@
+    %% Create schema:
+    Schema = #{
+        bits_per_wildcard_level => BitsPerTopicLevel,
+        topic_index_bytes => TopicIndexBytes,
+        ts_bits => 64,
+        ts_offset_bits => TSOffsetBits
+    },
+    {Schema, [{DataCFName, DataCFHandle}, {TrieCFName, TrieCFHandle}]}.
+
+-spec open(
+    emqx_ds_replication_layer:shard_id(),
+    rocksdb:db_handle(),
+    emqx_ds_storage_layer:gen_id(),
+    emqx_ds_storage_layer:cf_refs(),
+    schema()
+) ->
+    s().
+open(_Shard, DBHandle, GenId, CFRefs, Schema) ->
+    #{
+        bits_per_wildcard_level := BitsPerTopicLevel,
+        topic_index_bytes := TopicIndexBytes,
+        ts_bits := TSBits,
+        ts_offset_bits := TSOffsetBits
+    } = Schema,
+    {_, DataCF} = lists:keyfind(data_cf(GenId), 1, CFRefs),
+    {_, TrieCF} = lists:keyfind(trie_cf(GenId), 1, CFRefs),
+    Trie = restore_trie(TopicIndexBytes, DBHandle, TrieCF),
+    %% If user's topics have more than learned 10 wildcard levels
+    %% (more than 2, really), then it's total carnage; learned topic
+    %% structure won't help.
+    MaxWildcardLevels = 10,
+    KeymapperCache = array:from_list(
+        [
+            make_keymapper(TopicIndexBytes, BitsPerTopicLevel, TSBits, TSOffsetBits, N)
+         || N <- lists:seq(0, MaxWildcardLevels)
+        ]
+    ),
+    #s{db = DBHandle, data = DataCF, trie = Trie, keymappers = KeymapperCache}.
+
+-spec store_batch(
+    emqx_ds_replication_layer:shard_id(), s(), [emqx_types:message()], emqx_ds:message_store_opts()
+) ->
+    emqx_ds:store_batch_result().
+store_batch(_ShardId, S = #s{db = DB, data = Data}, Messages, _Options) ->
+    lists:foreach(
+        fun(Msg) ->
+            {Key, _} = make_key(S, Msg),
+            Val = serialize(Msg),
+            rocksdb:put(DB, Data, Key, Val, [])
+        end,
+        Messages
+    ).
+
+get_streams(_Shard, #s{trie = Trie}, TopicFilter, _StartTime) ->
+    Indexes = emqx_ds_lts:match_topics(Trie, TopicFilter),
+    [#stream{storage_key = I} || I <- Indexes].
+
+make_iterator(_Shard, _Data, #stream{storage_key = StorageKey}, TopicFilter, StartTime) ->
+    %% Note: it's a good idea to keep the iterator structure lean,
+    %% since it can be stored on a remote node that could update its
+    %% code independently from us.
+    {ok, #it{
+        topic_filter = TopicFilter,
+        start_time = StartTime,
+        storage_key = StorageKey
+    }}.
+
+next(_Shard, #s{db = DB, data = CF, keymappers = Keymappers}, It0, BatchSize) ->
+    #it{
+        start_time = StartTime,
+        storage_key = StorageKey
+    } = It0,
+    %% Make filter:
+    {TopicIndex, Varying} = StorageKey,
+    Inequations = [
+        {'=', TopicIndex},
+        {'>=', StartTime}
+        | lists:map(
+            fun
+                ('+') ->
+                    any;
+                (TopicLevel) when is_binary(TopicLevel) ->
+                    {'=', hash_topic_level(TopicLevel)}
+            end,
+            Varying
```

Comment:
Q: Should we compress topic levels past some threshold into a single hash to ensure that we have a hard limit on key size?

Reply:
I'm not sure I understood the idea: is it about replacing `hash(Varying1) ++ hash(Varying2)` with `hash(Varying1 ++ Varying2)`? This will make replays requireing strict match on `Varying1` and `+` on Varying2 less efficient.

Reply:
I'm slightly concerned about the fact that the key size currently is proportional to the number of topic levels, and number of topic levels is essentially limited only by the maximum length of an MQTT UTF-8 string. So it seems natural to have some hard limit here, so that the key contains no more than, say, 8 hashes.
```
[L1, L2, L3, L4, L5, L6, L7 | VaryingRest] = Varying
```

Admittedly, it should be hard to achieve such number of different varying topic levels, though I can imagine we'd want to have threshold = 0 for high topic levels in the LTS trie to contain its uncontrolled growth, which can lead to a lot of varying topic levels in some situations.

Reply:
https://emqx.atlassian.net/browse/EMQX-11274